Update changelog for 0.3.2 (#355)
PennyLaneAI/catalyst: doc/changelog.md

# Release 0.3.2

> <h3>New features</h3>

Reviewer: Just a reminder to include PR #347 changelog.

Author: done!

* The experimental AutoGraph feature now supports Python `while` loops, allowing native Python loops
  to be captured and compiled with Catalyst.
  [(#318)](https://github.com/PennyLaneAI/catalyst/pull/318)

  ```python

>   dev = qml.device("lightning.qubit", wires=4)

Author: I moved this to a separate call, as there is still the issue that memory allocation happens on device instantiation, and until this is fixed, want to avoid repeating device instantiation in the QNode decorator becoming a user pattern

  @qjit(autograph=True)
  @qml.qnode(dev)
  def circuit(n: int, x: float):
      i = 0

      while i < n:
          qml.RX(x, wires=i)
          i += 1

      return qml.expval(qml.PauliZ(0))
  ```

  ```pycon
  >>> circuit(4, 0.32)
  array(0.94923542)
  ```

>   This feature extends the existing AutoGraph support for Python `for` loops and if` statements

Reviewer: ```suggestion
  This feature extends the existing AutoGraph support for Python `for` loops and `if` statements
```

  introduced in v0.3. Note that TensorFlow must be installed for AutoGraph support.

  For more details, please see the
  [AutoGraph guide](https://docs.pennylane.ai/projects/catalyst/en/stable/dev/autograph.html).

> * In addition to loops and conditional branches, AutoGraph now supports native Python `and`, `or`

Reviewer: We need to add these new supports to the AutoGaraph guide page

Reviewer: I didn't review these PRs, otherwise I'd recommend it there

Author: Doing that now :)

  and `not` operators in Boolean expressions.
  [(#325)](https://github.com/PennyLaneAI/catalyst/pull/325)

  ```python
  dev = qml.device("lightning.qubit", wires=1)

  @qjit(autograph=True)
  @qml.qnode(dev)
  def circuit(x: float):

      if x >= 0 and x < jnp.pi:
          qml.RX(x, wires=0)

      return qml.probs()
  ```

  ```pycon
  >>> circuit(0.43)
  array([0.95448287, 0.04551713])
  >>> circuit(4.54)
  array([1., 0.])
  ```

  Note that Boolean expressions will only be captured by AutoGraph if at least
  one of the arguments is a numeric object.

> * It is now possible to dynamically load third-party Catalyst compatible devices directly
>   into a pre-installed Catalyst runtime.

Reviewer: It turned out we need to do more work to get this working on macOS :( 

```suggestion
* It is now possible to dynamically load third-party Catalyst compatible devices directly
  into a pre-installed Catalyst runtime on Linux.
```

Author: oh really? Not MacOS?

Reviewer: No, there are some API incompatibilities :(

  [(#327)](https://github.com/PennyLaneAI/catalyst/pull/327)

  To take advantage of this, third-party devices must implement the `Catalyst::Runtime::QuantumDevice`
  interface, in addition to defining the following method:

  ```cpp
  extern "C" Catalyst::Runtime::QuantumDevice*
  getCustomDevice() { return new CustomDevice(); }
  ```

  This support can also be integrated into existing PennyLane Python devices that inherit from
  the `QuantumDevice` class, by defining the `get_c_interface` static method.

  For more details, see the
  [custom devices documentation](https://docs.pennylane.ai/projects/catalyst/en/stable/dev/custom_devices.html).

> <h3>Improvements</h3>

Reviewer 2: Could you add https://github.com/PennyLaneAI/catalyst/pull/307 in improvements with the text. ```The scatter operation in the Catalyst dialect now uses a SCF for loop to avoid ballooning the compiled code.```

Author: Oh, I remember that used to be there, I wonder what happened to it

  ```python
  dev = qml.device("lightning.qubit", wires=4)

  @qjit(autograph=True)
  @qml.qnode(dev)
  def circuit(n: int, x: float):
      i = 0

      while i < n:
          qml.RX(x, wires=i)
          i += 1

      return qml.expval(qml.PauliZ(0))
  ```

  ```pycon
  >>> circuit(4, 0.32)
  array(0.94923542)
  ```

  This feature extends the existing AutoGraph support for Python `for` loops and `if` statements
  introduced in v0.3. Note that TensorFlow must be installed for AutoGraph support.

  For more details, please see the
  [AutoGraph guide](https://docs.pennylane.ai/projects/catalyst/en/stable/dev/autograph.html).

* In addition to loops and conditional branches, AutoGraph now supports native Python `and`, `or`
  and `not` operators in Boolean expressions.
  [(#325)](https://github.com/PennyLaneAI/catalyst/pull/325)

  ```python
  dev = qml.device("lightning.qubit", wires=1)

  @qjit(autograph=True)
  @qml.qnode(dev)
  def circuit(x: float):

      if x >= 0 and x < jnp.pi:
          qml.RX(x, wires=0)

      return qml.probs()
  ```

  ```pycon
  >>> circuit(0.43)
  array([0.95448287, 0.04551713])
  >>> circuit(4.54)
  array([1., 0.])
  ```

>   Note that Boolean expressions will only be captured by AutoGraph if at least
>   one of the arguments is a numeric object.

Reviewer 3: ```suggestion
  Note that logical Boolean operators will only be captured by AutoGraph if all operands
  are dynamic variables (e.g. a measurement result or function argument). For other use
  cases, it is recommended to use the `jax.numpy.logical_*` set of functions where
  appropriate.
```

  ```python
  @qjit(autograph=True)
  def func(i: int, f: float):

      if i < 3:
          i = i
      else:
          i = f

      return i
  ```

  ```pycon
  >>> func(1, 4.0)
  array(1.0)
  ```

* AutoGraph now supports converting functions even when they are invoked through functional wrappers such
  as `adjoint`, `ctrl`, `grad`, `jacobian`, etc.
  [(#336)](https://github.com/PennyLaneAI/catalyst/pull/336)

  For example, the following should now succeed:

  ```python
  def inner(n):
    for i in range(n):
      qml.T(i)

  @qjit(autograph=True)
  @qml.qnode(dev)
  def f(n: int):
      adjoint(inner)(n)
      return qml.state()
  ```

* To prepare for frontend integration with PennyLane, the `pennylane.compilers` entry points interface
  has added to Catalyst.
  [(#331)](https://github.com/PennyLaneAI/catalyst/pull/331)

  For any compiler packages seeking to be registered in PennyLane, the `entry_points`
  metadata under the the group name `pennylane.compilers` must be added, with the following entry points:

  - `context`: Path to the compilation evaluation context manager. This context manager should have

>     the method `context.is_tracing()``, which returns True if called within a program that is being

Reviewer: ```suggestion
    the method `context.is_tracing()`, which returns True if called within a program that is being
```

  array(1.0)
  ```

>   Automatic type promotion across conditional branches also now works with AutoGraph:

Reviewer 3: Is this necessary for the changelog, seems like it is just an implicit effect of the above no?

Author: It is an implicit effect of the above, but wanted to make it clear (definitely feels like a 'free' additional improvement)

Reviewer 3: ```suggestion
  Automatic type promotion across conditional branches also works with AutoGraph:
```

  ```python
  @qjit(autograph=True)
  def func(i: int, f: float):

      if i < 3:
          i = i
      else:
          i = f

      return i
  ```

  ```pycon
  >>> func(1, 4.0)
  array(1.0)
  ```

* AutoGraph now supports converting functions even when they are invoked through functional wrappers such
  as `adjoint`, `ctrl`, `grad`, `jacobian`, etc.
  [(#336)](https://github.com/PennyLaneAI/catalyst/pull/336)

  For example, the following should now succeed:

  ```python
  def inner(n):
    for i in range(n):
      qml.T(i)

  @qjit(autograph=True)
  @qml.qnode(dev)
  def f(n: int):
      adjoint(inner)(n)
      return qml.state()
  ```

> * To prepare for frontend integration with PennyLane, the `pennylane.compilers` entry points interface
>   has added to Catalyst.

Reviewer 3: ```suggestion
* To prepare for Catalyst's frontend being integrated with PennyLane, the appropriate plugin entry point
  interface has been added to Catalyst.
```

Author: 👍

  [(#331)](https://github.com/PennyLaneAI/catalyst/pull/331)

  For any compiler packages seeking to be registered in PennyLane, the `entry_points`
  metadata under the the group name `pennylane.compilers` must be added, with the following entry points:

  - `context`: Path to the compilation evaluation context manager. This context manager should have
    the method `context.is_tracing()`, which returns True if called within a program that is being
    traced or captured.

  - `ops`: Path to the compiler operations module. This operations module may contain compiler
    specific versions of PennyLane operations. Within a JIT context, PennyLane operations may
    dispatch to these.

  - `qjit`: Path to the JIT compiler decorator provided by the compiler. This decorator should have

>     the signature `qjit(fn, *args, **kwargs)`, where `fn` is the function to be compiled.
> * The compiler driver diagnostic output has been improved, and now includes failing IR as well as

Reviewer 3: ```suggestion
    the signature `qjit(fn, *args, **kwargs)`, where `fn` is the function to be compiled.
  
* The compiler driver diagnostic output has been improved, and now includes failing IR as well as
```

>   names of failing passes.

Reviewer 3: ```suggestion
  the names of failing passes.
```

  [(#349)](https://github.com/PennyLaneAI/catalyst/pull/349)

> * The scatter operation in the Catalyst dialect now uses a SCF for loop to avoid ballooning

Reviewer 3: ```suggestion
* The scatter operation in the Catalyst dialect now uses an SCF for loop to avoid ballooning
```

  the compiled code.
  [(#307)](https://github.com/PennyLaneAI/catalyst/pull/307)

* The `CopyGlobalMemRefPass` pass of our MLIR processing pipeline now supports
  dynamically shaped arrays.
  [(#348)](https://github.com/PennyLaneAI/catalyst/pull/348)

<h3>Breaking changes</h3>

> * The axis ordering for `catalyst.jacobian` is updated to match `jax.jacobian`. Assume we have

Reviewer 3: ```suggestion
* The axis ordering for `catalyst.jacobian` is updated to match `jax.jacobian`. Assuming we have
```

  parameters of shape `[a,b]` and results of shape `[c,d]`, the returned Jacobian will now have
  shape `[c, d, a, b]` instead of `[a, b, c, d]`.
  [(#283)](https://github.com/PennyLaneAI/catalyst/pull/283)

<h3>Bug fixes</h3>

> * Fixes the issue with missing `CFP_t` in `StateVectorLQubitDynamic` when building against the
>   master branch of PennyLane-Lightning. This issue was introduced in [(#499)](https://github.com/PennyLaneAI/pennylane-lightning/pull/499).

Reviewer 3: ```suggestion
* An upstream change in the PennyLane-Lightning project was addressed to prevent compilation issues
  in the `StateVectorLQubitDynamic` class in the runtime.
  The issue was introduced in [#499](https://github.com/PennyLaneAI/pennylane-lightning/pull/499).
```

Author: Thank you! This is much clearer

  [(#322)](https://github.com/PennyLaneAI/catalyst/pull/322)

* The `requirements.txt` file to build Catalyst from source has been updated with a minimum pip
  version, `>=22.3`. Previous versions of pip are unable to perform editable installs when the
  system-wide site-packages are read-only, even when the `--user` flag is provided.
  [(#311)](https://github.com/PennyLaneAI/catalyst/pull/311)

* The frontend has been updated to make it compatible with PennyLane `MeasurementProcess` objects
  now being PyTrees in PennyLane version 0.33.
  [(#315)](https://github.com/PennyLaneAI/catalyst/pull/315)

* The Catalyst utility dialect is now included in the Catalyst MLIR C-API.

>   [(#345)](https://github.com/PennyLaneAI/catalyst/pull/345)

Reviewer 3: This is an improvement rather than a bug fix :)

Author: Oops. Definitely move it up in that case!

  shape `[c, d, a, b]` instead of `[a, b, c, d]`.
  [(#283)](https://github.com/PennyLaneAI/catalyst/pull/283)

<h3>Bug fixes</h3>

* Fixes the issue with missing `CFP_t` in `StateVectorLQubitDynamic` when building against the
  master branch of PennyLane-Lightning. This issue was introduced in [(#499)](https://github.com/PennyLaneAI/pennylane-lightning/pull/499).
  [(#322)](https://github.com/PennyLaneAI/catalyst/pull/322)

* The `requirements.txt` file to build Catalyst from source has been updated with a minimum pip
  version, `>=22.3`. Previous versions of pip are unable to perform editable installs when the
  system-wide site-packages are read-only, even when the `--user` flag is provided.
  [(#311)](https://github.com/PennyLaneAI/catalyst/pull/311)

* The frontend has been updated to make it compatible with PennyLane `MeasurementProcess` objects
  now being PyTrees in PennyLane version 0.33.
  [(#315)](https://github.com/PennyLaneAI/catalyst/pull/315)

* The Catalyst utility dialect is now included in the Catalyst MLIR C-API.
  [(#345)](https://github.com/PennyLaneAI/catalyst/pull/345)

> <h3>Contributors</h3>

Reviewer 2: You should add yourself and @erick-xanadu to the contributors.

>   - `qjit`: Path to the JIT compiler decorator provided by the compiler. This decorator should have
>     the signature `qjit(fn, *args, **kwargs)`, where fn is the function to be compiled.
> 

Reviewer: ```suggestion
  - `qjit`: Path to the JIT compiler decorator provided by the compiler. This decorator should have
    the signature `qjit(fn, *args, **kwargs)`, where `fn` is the function to be compiled.
```
